lib/src/boundimage.rs

```diff
@@ -139,29 +139,36 @@ struct BoundImage {
 
 impl BoundImage {
     fn new(image: String, auth_file: Option<String>) -> Result<BoundImage> {
-        validate_spec_value(&image).context("Invalid image value")?;
+        let image = parse_spec_value(&image).context("Invalid image value")?;
 
-        if let Some(auth_file) = &auth_file {
-            validate_spec_value(auth_file).context("Invalid auth_file value")?;
-        }
+        let auth_file = if let Some(auth_file) = &auth_file {
+            Some(parse_spec_value(auth_file).context("Invalid auth_file value")?)
+        } else {
+            None
+        };
 
         Ok(BoundImage { image, auth_file })
     }
 }
 
-fn validate_spec_value(value: &String) -> Result<()> {
-    let mut number_of_percents = 0;
+fn parse_spec_value(value: &String) -> Result<String> {
+    let mut parsed_value = String::new();
+    let mut prev_is_percent = false;
```

Comment:
I think this is almost right - if I'm thinking through this correctly there's just a corner case where if the string ends in a trailing `%` I believe we will silently swallow it and not error out.

How about this code:

```
fn parse_spec_value(value: &str) -> Result<String> {
    let mut it = s.chars();
    let mut ret = String::new();
    while let Some(c) = it.next() {
        if c != '%' {
            ret.push(c);
            continue;
        }
        let c = it.next().ok_or_else(|| anyhow!("Unterminated %"))?;
        match c {
            '%' => { ret.push('%'); },
            _ => anyhow::bail!("Systemd specifiers are not supported by bound bootc images: {value}");
       }
    Ok(r)
}
```
or so?

Reply:
ah thanks! I was struggling with iterators but this clears it up. I was missing that it.next() within the while will advance the iterator.